Fix correct parsing of collections in AmazonLambdaRecorder

hamburml opened the pull request:
test to make sure that the objectMapper parses the Input correct.

@patriot1burke thx for the hint.

I also adjusted the objectWriter but I am not sure if this is really necessary.

- Fixes: https://github.com/quarkusio/quarkus/issues/40349

hamburml committed:
Fix correct parsing of collections in AmazonLambdaRecorder with CollectionInputReader
add test for abstractHandler
add test for collection input and collection output
rename tests for clarification what they are for

gsmet commented on extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java:
```diff
@@ -53,12 +55,14 @@ static void initializeHandlerClass(Class<? extends RequestHandler<?, ?>> handler
         if (parameterType.equals(S3Event.class)) {
             objectReader = new S3EventInputReader(objectMapper);
         } else {
-            objectReader = new JacksonInputReader(objectMapper.readerFor(parameterType));
+            objectReader = new JacksonInputReader(objectMapper.readerFor(javaType));
         }
-        objectWriter = new JacksonOutputWriter(objectMapper.writerFor(handlerMethod.getReturnType()));
+        objectWriter = new JacksonOutputWriter(objectMapper.writerFor(javaType));
```
I'm surprised by this change? It used to use the return type and now AFAICS it uses the parameter type? Is it expected?

hamburml replied:
You are right. To be honest I simply adopted @patriot1burke comment in the issue and didnt gave it more thought. Just wanted to start the process of fixing this.

gsmet reviewed (changes requested):
Thanks for the PR! I spotted something that looks a bit odd to me.

Also the tests are not passing.

gsmet commented on extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java:
```diff
@@ -53,12 +55,17 @@ static void initializeHandlerClass(Class<? extends RequestHandler<?, ?>> handler
         ObjectMapper objectMapper = AmazonLambdaMapperRecorder.objectMapper;
         Method handlerMethod = discoverHandlerMethod(handlerClass);
         Class<?> parameterType = handlerMethod.getParameterTypes()[0];
+        Type genericParameterType = handlerMethod.getGenericParameterTypes()[0];
+        JavaType constructParameterType = objectMapper.getTypeFactory().constructType(genericParameterType);
+        Type genericReturnType = handlerMethod.getGenericReturnType();
+        JavaType constructReturnType = objectMapper.getTypeFactory().constructType(genericReturnType);
+
```
I think you'd rather move these lines just where they are needed. As typically in the `S3Event` case, some of this is useless.

gsmet commented on extensions/amazon-lambda/deployment/src/test/java/io/quarkus/amazon/lambda/deployment/testing/LambdaHandlerET.java:
```diff
@@ -5,18 +5,14 @@
 
 import org.junit.jupiter.api.Test;
 
-import io.quarkus.test.junit.QuarkusTest;
-
-@QuarkusTest
```
Oh, sorry, that I think we should keep.

gsmet reviewed (approved):
CI is happy so let's merge this.

Thanks for your contribution!

gsmet commented:
@patriot1burke are you happy with it? If so, we can merge.

gsmet commented:
@patriot1burke which reviews? I don't see any in this PR? Could you point me to where your comments are?

geoand commented:
Maybe the review was not posted by accident (the GH UI tends to make this extremely easy to happen)?

hamburml commented:
Maybe he means this comment in the linked Issue https://github.com/quarkusio/quarkus/issues/40349#issuecomment-2085262068 
To be honest I never had issues with the output of the lambda, so I did not do the same to the objectWriter.

patriot1burke commented on extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java:
```diff
@@ -53,11 +55,15 @@ static void initializeHandlerClass(Class<? extends RequestHandler<?, ?>> handler
             objectReader = new JacksonInputReader(objectMapper.readerFor(parameterType));
         }
+
         objectWriter = new JacksonOutputWriter(objectMapper.writerFor(handlerMethod.getReturnType()));
```
Not seeing these?

hamburml replied:
pls check new commit, i didnt need to to anything with the returnType and it worked for a collection.

patriot1burke reviewed (commented):
See my reviews now?

patriot1burke commented:
> Maybe he means this comment in the linked Issue [#40349 (comment)](https://github.com/quarkusio/quarkus/issues/40349#issuecomment-2085262068) To be honest I never had issues with the output of the lambda, so I did not do the same to the objectWriter.

That is one of the comments.  This PR is incomplete without supporting generic return type.

hamburml commented:
@patriot1burke Now your comments are visible. Maybe tomorrow I will find some time. Not sure. Thanks.

hamburml commented:
@patriot1burke I just pushed a new Test. Now I have a `RequestHandler<List<Person>, List<Person>>` and a `RequestHandler<List<String>, List<Person>>` lambda handler. They work out of the box, didn't need to change the writer. Any advice is appreciated, I do not know what I should do.

hamburml commented on extensions/amazon-lambda/runtime/src/main/java/io/quarkus/amazon/lambda/runtime/AmazonLambdaRecorder.java:
```diff
@@ -53,11 +55,15 @@ static void initializeHandlerClass(Class<? extends RequestHandler<?, ?>> handler
         if (parameterType.equals(S3Event.class)) {
             objectReader = new S3EventInputReader(objectMapper);
+        } else if (Collection.class.isAssignableFrom(parameterType)) {
+            objectReader = new CollectionInputReader<>(objectMapper, handlerMethod);
```
IMHO, this looks consistent with the other code. I can change it if necessary.

hamburml commented:
@patriot1burke I do not give up! I just renamed the tests and added new ones.
We now have a Test for an handleRequest which comes from an abstract class. 
We now have a Test where handleRequest gets as Input a Collection and as output a Collection.
The tests are green on my end. Please take a look.

hamburml commented:
@patriot1burke anything pls?

patriot1burke commented:
There's a conflict now with main.  Need to resolve conflicts

hamburml commented:
> There's a conflict now with main. Need to resolve conflicts

I just rebased. I cant see any conflicts. Am I missing something here?

hamburml commented:
This is really strange. Under GitHub actions I see the workflow run and it looks like it does not run. But I do not know why.

patriot1burke commented:
I don't see the conflict anymore

gsmet commented:
And merged! Thanks a lot for your efforts!

hamburml commented:
@gsmet could we backport this to current LTS?

geoand commented:
+1 on backporting, but we'll also need https://github.com/quarkusio/quarkus/pull/41896 because the test that was added with this is flaky
